### PR TITLE
support faster rcnn

### PR DESCRIPTION
**Pull request by wenyushi**
The push will affect:
1. dlpy/applications.py: added a helper function to get parameter of a layer which makes it easier to manage and expand our code; supported Faster RCNN model.
2. dlpy/utils.py: add a function, plot_anchors, to visualize proposed anchor boxes
3. dlpy/tests/test_applications.py: a simple test case covering Faster RCNN.
4. dlpy/tests/test_utils.py: added a test case to cover plot_anchors()

### CODE REVIEW

File: dlpy/applications.py [hunk 1]
```diff
@@ -20,17 +20,44 @@
...
+import six
 
 from .sequential import Sequential
 from .blocks import ResBlockBN, ResBlock_Caffe, DenseNetBlock, Bidirectional
 from .caffe_models import (model_vgg16, model_vgg19, model_resnet50,
                            model_resnet101, model_resnet152)
 from .keras_models import model_inceptionv3
-from .layers import (InputLayer, Conv2d, Pooling, Dense, BN, OutputLayer, Detection, Concat, Reshape, Recurrent,
-                     GlobalAveragePooling2D, GroupConv2d, Conv2DTranspose)
+from .layers import (Input, InputLayer, Conv2d, Pooling, Dense, BN, OutputLayer, Detection, Concat, Reshape, Recurrent,
```
Comment (ethem-kinginthenorth) [inline dlpy/applications.py]: I am assuming we are not using them here.

File: dlpy/applications.py [hunk 2]
```diff
@@ -20,17 +20,44 @@
...
+                       'dropout', 'random_flip', 'random_crop', 'random_mutation', 'norm_stds']
+rpn_layer_options = ['anchor_ratio', 'anchor_scale', 'anchor_num_to_sample', 'base_anchor_size',
+                     'coord_type', 'do_RPN_only', 'max_label_per_image', 'proposed_roi_num_score',
+                     'proposed_roi_num_train', 'roi_train_sample_num']
+fast_rcnn_options = ['detection_threshold', 'max_label_per_image', 'max_object_num', 'nms_iou_threshold']
+
+
+def _get_layer_options(layer_options, local_options):
+    """
+    Get parameters belonging to certain type of layer.
```
Comment (ethem-kinginthenorth) [inline dlpy/applications.py]: to "a" certain

File: dlpy/applications.py [hunk 3]
```diff
@@ -20,17 +20,44 @@
...
 
+input_layer_options = ['n_channels', 'width', 'height', 'nominals', 'std', 'scale', 'offsets',
+                       'dropout', 'random_flip', 'random_crop', 'random_mutation', 'norm_stds']
+rpn_layer_options = ['anchor_ratio', 'anchor_scale', 'anchor_num_to_sample', 'base_anchor_size',
+                     'coord_type', 'do_RPN_only', 'max_label_per_image', 'proposed_roi_num_score',
+                     'proposed_roi_num_train', 'roi_train_sample_num']
+fast_rcnn_options = ['detection_threshold', 'max_label_per_image', 'max_object_num', 'nms_iou_threshold']
+
+
+def _get_layer_options(layer_options, local_options):
```
Comment (ethem-kinginthenorth) [inline dlpy/applications.py]: please add some comments on what you are doing here.

File: dlpy/applications.py [hunk 4]
```diff
@@ -4001,3 +4028,147 @@ def InceptionV3(conn, model_table='InceptionV3',
...
+    Parameters
+    ----------
+    conn : CAS
+        Specifies the connection of the CAS connection.
+    model_table : string, optional
+        Specifies the name of CAS table to store the model.
+    n_channels : int, optional
+        Specifies the number of the channels (i.e., depth) of the input layer.
+        Default: 3
+    width : int, optional
```
Comment (ethem-kinginthenorth) [inline dlpy/applications.py]: really, we have a skewed image :)

File: dlpy/applications.py [hunk 5]
```diff
@@ -4001,3 +4028,147 @@ def InceptionV3(conn, model_table='InceptionV3',
...
+    Returns
+    -------
+    :class:`Sequential`
+
+    References
+    ----------
+    https://arxiv.org/abs/1506.01497
+
+    '''
+    num_anchors = len(anchor_ratio) * len(anchor_scale)
```
Comment (ethem-kinginthenorth) [inline dlpy/applications.py]: please add some comments (e.g., calculating parameters here, building mode here, and so on)

File: dlpy/tests/test_applications.py [hunk 6]
```diff
@@ -566,3 +566,8 @@ def test_speech_recognition_1(self):
         from dlpy.applications import SpeechRecognition
         model = SpeechRecognition(self.s, n_blocks=1)
         model.print_summary()
+
```
Comment (ethem-kinginthenorth) [inline dlpy/tests/test_applications.py]: can you please also add a couple of more tests using different parameters than the default ones

File: dlpy/utils.py [hunk 7]
```diff
@@ -1444,6 +1445,60 @@ def display_object_detections(conn, table, coord_type, max_objects=10,
...
+
+    '''
+    color_map = ['b', 'g', 'r', 'c', 'm', 'y']
+    img_height = image_size[0]
+    img_width = image_size[1]
+    anchors = []
+    max_anchor_height = image_size[0]
+    max_anchor_width = image_size[1]
+
+    for ratio in anchor_ratio:
```
Comment (ethem-kinginthenorth) [inline dlpy/utils.py]: comment please

File: dlpy/utils.py [hunk 8]
```diff
@@ -1444,6 +1445,60 @@ def display_object_detections(conn, table, coord_type, max_objects=10,
...
+        max_anchor_width = max(max_anchor_width, an[1])
+    fig, ax = plt.subplots(1, figsize = fig_size)
+    plt.xticks([]), plt.yticks([])
+    background = np.tile((255, 255, 255), (int(max_anchor_height), int(max_anchor_width), 1))
+    image_region = (int((max_anchor_height - img_height) / 2), int((max_anchor_height + img_height) / 2),
+                    int((max_anchor_width - img_width) / 2), int((max_anchor_width + img_width) / 2))
+    background[image_region[0]: image_region[1], image_region[2]: image_region[3], :] = np.array((244, 203, 66))
+    ax.imshow(background)
+
+    for i, anchor in enumerate(anchors):
```
Comment (ethem-kinginthenorth) [inline dlpy/utils.py]: comment please

File: dlpy/applications.py (hunk 1 above)
Comment (wenyushi) [reply dlpy/applications.py]: Yes, GroupConv2d, Conv2DTranspose will come in the next push related to UNet and MobileNet

File: dlpy/applications.py (hunk 4 above)
Comment (wenyushi) [reply dlpy/applications.py]: Xindian gave me a VOC dataset that is in the size.

**Pull request merged**